src/main/kotlin/com/shortener/service/UrlServiceImpl.kt

```diff
@@ -0,0 +1,101 @@
+package com.shortener.service
+
+import com.shortener.domain.UrlEntry
+import com.shortener.domain.UrlEntryRepository
+import com.shortener.dto.UrlShortenRequest
+import com.shortener.dto.UrlShortenResponse
+import com.shortener.exception.InvalidInputUrl
+import com.shortener.utils.IdEncoder
+import org.apache.commons.validator.routines.UrlValidator
+import org.apache.commons.validator.routines.UrlValidator.ALLOW_LOCAL_URLS
+import org.springframework.stereotype.Service
+import org.springframework.web.servlet.support.ServletUriComponentsBuilder
+import java.time.LocalDateTime
+import java.util.*
+
+private const val HTTP_SCHEMA = "http"
+private const val DEFAULT_HTTPS_SCHEMA = "https"
+private const val SCHEMA_SEPARATOR = "://"
+private const val URL_EXPIRES_DAYS_NO_USER = 7L
+
+@Service
+class UrlServiceImpl(private val urlEntryRepository: UrlEntryRepository, private val idEncoder: IdEncoder) :
+    UrlService {
+
+    override fun shortenUrl(request: UrlShortenRequest): UrlShortenResponse {
+        val fixedRequestUrl = addDefaultSchemaIfMissing(request.longUrl)
+
+        if (!isValidUrl(fixedRequestUrl)) {
+            throw InvalidInputUrl("The URL you provided is invalid")
+        }
+
+        val urlEntry = UrlEntry()
+        urlEntry.setLongUrl(fixedRequestUrl)
+        urlEntry.setCreatedAt(LocalDateTime.now())
+        urlEntry.setExpiresAt(urlEntry.getCreatedAt()?.plusHours(getNoUserUrlExpiresDays()))
+
+        val persistedUrlEntryId = urlEntryRepository.save(urlEntry).getId()
+        val encodedUrlEntryId = idEncoder.encode(persistedUrlEntryId!!)
+        val url = addBaseUrlToEncodedSequence(encodedUrlEntryId)
+        return UrlShortenResponse(url)
+    }
+
+    override fun decodeUrl(encodedSequence: String): String {
```

Comment:
This thing looks bad performance wise imo: 

If i were to read the same short url (into a long url) `1 000 000 000 000` times, we would have to call the decode function every time.

Why not persist the shortened URL with every entry as well? 

Right now we hit the database and also hit the CPU.

Secondary to this: what happens when we change the url encoder? My gut feeling tells me we will not be able to find the long url by the short one, as the decoder will not return the correct data. This is very wrong ❗

Reply:
I dont really care about performance or if the application is correct or not, but something cool that you might want to look into is Hazelcast to cache and avoid rendundant database calls.

P.S: Hazelcast is really easy to add and you dont need a different cluster for it, the node will run in the same JVM as your application.

Reply:
Thanks for the caching suggestions. I am currently working on it, in different branch. Decided to use `redis`